compiler/src/iree/compiler/Codegen/LLVMGPU/ROCDLBufferInstructionsOptimization.cpp

```diff
@@ -0,0 +1,103 @@
+// Copyright 2025 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/Codegen/LLVMGPU/Passes.h"
+#include "iree/compiler/Codegen/Utils/GPUUtils.h"
+#include "iree/compiler/Codegen/Utils/Utils.h"
+#include "llvm/Support/Debug.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#define DEBUG_TYPE "iree-codegen-rocdl-buffer-instructions-optimization"
+#define DBGS() (llvm::dbgs() << "[" DEBUG_TYPE "]: ")
+#define LDBG(X) LLVM_DEBUG(DBGS() << X << "\n")
+
+namespace mlir::iree_compiler {
+
+#define GEN_PASS_DEF_ROCDLBUFFERINSTRUCTIONSOPTIMIZATIONPASS
+#include "iree/compiler/Codegen/LLVMGPU/ROCDLPasses.h.inc"
+
+namespace {
+
+Value createi1And(Location loc, ArrayRef<Value> values, OpBuilder &builder) {
+  Value base =
+      builder.create<arith::IndexCastUIOp>(loc, builder.getI1Type(), values[0]);
+  for (auto value : values.drop_front()) {
+    Value rhs =
+        builder.create<arith::IndexCastUIOp>(loc, builder.getI1Type(), value);
+    base = builder.create<arith::AndIOp>(loc, base, rhs);
+  }
+  return base;
+}
+
+class SimplyMaskVectorTransferRead
+    : public OpRewritePattern<vector::CreateMaskOp> {
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(vector::CreateMaskOp maskOp,
```

Comment:
I don't think this'll be correct when dealing with underaligned vectors of sub-32bit types - see all the stuff @jerryyin's had to do upstream

Reply:
Good point, it didn't occur to me when we are talking offline. Yes, in order for this pass to work correctly, it needs to deal with sub-word aligned problem. 

I think we have two options then:
 1. Add this pass on amdgpu pattern, but you'll have to deal with extract_slice of mask to backtrack to the real mask, then optimize. This way you can re-use the `scf.if` logic already populated from that pass.
 2. We'd need to move the amdgpu scf.if population into a standalone utility. Then we'd have both the amdgpu pass as well as this one to take advantage of the common utility and populate the `scf.if` condition check.

Option 2 is a more cleaner approach compared with option 1 I think.

Reply:
Let me add some tests today to make the scope of this pattern clear. It should never trigger on unaligned subword situation. It's for when the mask is all ones or all zeros. So if the mask is all ones. It becomes a regular transfer read. And rhat transfer read is guaranteed to be aligned as it's within bounds. If the mask is all zeros then it's irrelevant what the transfer read value is as long as it doesn't cause a crash as we will be using the padding value instead.

Reply:
I think the sub-word alignment problem goes orthogonal with the mask content. let's say we have <6xi8> buffer. Let's say you are trying to load a <8xi8> vector with mask all 1s. This will be a problem for the 5th and 6th i8 because those two bytes crossed the buffer length boundary and buffer load will give you all zeros, causing wrong results to be loaded.

Reply:
So yes, realistically the last two mask bits are zero, but that's byte 7 and 8 anyway (and is already out of bounds so doesn't matter). The problem is in the byte 5 and 6, regardless of the mask value, when you use buffer load, will always give you zeros.

Reply:
Yes, but this pattern won't happen in that realistic case. And if you set mask bits one for data that is out of bounds then we will have to fix that in our codegen I think.

Reply:
Note, for a <6xi8> buffer and doing a <8xi8> vector with mask of [6x1, 0, 0] is a possible pattern that I've seen a few times in codegen. It is possible for that to happen. From a codegen perspective, that's a valid codegen (considering that buffer load handles bound check on its own, so will avoid hardware faults).

Reply:
Ah that's right, your implementation only handles mask of all 1s. That indicate on all cases when it is k-aligned, it will not be handled by this pass. Good point!

Reply:
Yup! I think you mean k-unaligned will not be handled.

Reply:
So if your mask is either all 1s or all 0s, then your indices should either be the real indices or [dim(0), 0, ... 0]/[0, 0, ... size(tensor)] respectively 

And mark that OOB read as in-bounes because we don't want the transfer_read lowering to put in it statements

Reply:
Yes except the get all zeros part of mask is zero. You could be accessing something within the memref and still have mask zero if that's what codegen wants to do, in that case you will get real values but you want to use the padding values zero.

Reply:
That being said getting all zeros is also ok if it is indeed out of bounds. We won't really be looking at what we get.

Reply:
> So if your mask is either all 1s or all 0s, then your indices should either be the real indices or [dim(0), 0, ... 0]/[0, 0, ... size(tensor)] respectively
> 
> And mark that OOB read as in-bounes because we don't want the transfer_read lowering to put in it statements

Yes, I am marking the transfer_read in bounds.